pkg/internal/source/kind.go

```diff
@@ -24,72 +27,142 @@ type Kind struct {
+	// an error or the specified context is cancelled or expired.
+	if err := wait.PollImmediateUntilWithContext(ctx, 10*time.Second, func(pollCtx context.Context) (bool, error) {
+		// Lookup the Informer from the Cache and add an EventHandler which populates the Queue
+		i, lastErr = informerCache.GetInformer(pollCtx, resourceType)
+		if lastErr != nil {
+			kindMatchErr := &meta.NoKindMatchError{}
+			switch {
+			case errors.As(lastErr, &kindMatchErr):
+				// We got a NoKindMatchError, which means the kind is a CRD and it's not installed yet.
+				// We should retry until it's installed.
+				log.Error(lastErr, "if kind is a CRD, it should be installed before calling Start", "kind", kindMatchErr.GroupKind)
 				return false, nil // Retry.
+			case runtime.IsNotRegisteredError(lastErr):
+				// We got a IsNotRegisteredError, which means the kind is not registered to the Scheme.
+				// This is a programming error, so we should stop retrying and return the error.
+				return true, fmt.Errorf("kind must be registered to the Scheme: %w", lastErr)
+			default:
+				// We got an error that is not a NoKindMatchError or IsNotRegisteredError, so we should
+				// stop retrying and return the error.
+				return true, fmt.Errorf("failed to get informer from cache: %w", lastErr)
 			}
-			return true, nil
-		}); err != nil {
-			if lastErr != nil {
-				ks.started <- fmt.Errorf("failed to get informer from cache: %w", lastErr)
-				return
-			}
-			ks.started <- err
-			return
 		}
+		return true, nil
+	}); err != nil {
+		if lastErr != nil {
+			return nil, kerrors.NewAggregate([]error{err, fmt.Errorf("failed to get informer from cache: %w", lastErr)})
+		}
+		return nil, err
+	}
 
-		_, err := i.AddEventHandler(NewEventHandler(ctx, queue, handler, prct).HandlerFuncs())
-		if err != nil {
-			ks.started <- err
-			return
+	return i, nil
+}
+
+func (ks *Kind) registerEventHandler(ctx context.Context, eventHandler toolscache.ResourceEventHandlerFuncs) error {
+	informer, err := getInformer(ctx, ks.Cache, ks.Type)
+	if err != nil {
+		return err
+	}
+
+	err = func() error {
+		ks.mu.Lock()
+		defer ks.mu.Unlock()
+
+		// The source was canceled while we were getting the informer.
+		// We should not add the event handler or save the obtained informer.
+		if ks.canceled {
+			return nil
 		}
-		if !ks.Cache.WaitForCacheSync(ctx) {
-			// Would be great to return something more informative here
-			ks.started <- errors.New("cache did not sync")
+
+		// Save the informer so that we use it to unregister the event handler in Stop.
+		ks.informer = informer
+
+		// Register the event handler and save the registration so that we can unregister it in Stop.
+		registration, err := ks.informer.AddEventHandler(eventHandler)
+		if err != nil {
+			return err
 		}
-		close(ks.started)
+		ks.registration = registration
+
+		return nil
 	}()
+	if err != nil {
+		return err
+	}
+
+	// Wait for the cache to sync before adding the EventHandlers.
```

Comment:
```suggestion
	// Wait for the cache to sync.
```
nit: we already add the event handler before